### DISCUSSION TITLE
swapping ganged z for ganged y axis

### DISCUSSION THREAD

**Discussion by Bazmundi** [grblHAL/core#572] [Q&A] [answered]
[VEVOR](https://www.vevor.com.au/laser-engraving-machine-c_11141/vevor-desktop-laser-engraver-12-2-x11-8-large-engraving-area-5-5w-laser-power-p_010344972642) had a sale and I jumped on a 5.5W 12"x11" laser cutter, to scavange it and repurpose it as a solder paste "robot" (mostly for the challenge, DON'T JUDGE ME :blush: ).  Don't get exited, the sale is over but it was %60 off, so worth the hoot I thought.

I have a BTT SKR PICO 1.0 (on sale %70 off :grinning: ) and looking at how to hack gbrl HAL to do what I want. 

I am adding a single z axis to the rig, and using a spindle on/off to turn on a pneumatic syringe. 

The PICO is a ganged z axis? But the VEVOR laser cutter is actually a ganged y axis (or will be in the targeted setup).  

I am therefore assuming I need "lie" in btt_skr_pico_10_map.h, by swapping y and z pins? Straight forward.

Assuming that, then is there a means to pull that into the web build tool? Or (an I am getting that sinking feeling) I have to build it with the tool chain?

If it helps, I am setting up MSCode with the RPI PICO SDK etc.  Have found https://github.com/grblHAL/RP2040/discussions/82 (though that may be dated).  In the process of trying to set up for a build.  Wish me luck :wink:.

Has anyone tested the BTT SKR PICO build on the web builder at least please?

**Comment by terjeio** [comment]
Ganged (or auto squared) Y-axis can be selected in the Web Builder - by the _Y-axis_ drop down.

**Comment by Bazmundi** [comment]
Yes, but the point was the mismatch between the board (z ganged) and the
lazer cutter (y ganged). So, it will need hacking of the HAL code.

At the moment, I am seeing a fail to install the Pico SDK into MS Code, so
I am still working on a build.

**Reply by terjeio** [reply]
I do not understand the need for hacking...
If the laser cutter has the motors wired incorrectly it would be simpler to just swap the motors by plugging Z into Z and Y2 into the E0 socket?

**Comment by Bazmundi** [comment]
They are not wired incorrectly, there are 2 motors driving what would be the y axis on a 3D printer.  You otherwise have to swap all y and z commands in your g-code if you take your position.  Gotta script for that? 

I am adding a z-axis and driving a pneumatic pump to push solder past through a syringe.  The pump would be through an ON/OFF spindle option.

**Reply by terjeio** [reply]
> I am adding a z-axis and driving a pneumatic pump to push solder past through a syringe.  The pump would be through an ON/OFF spindle option.

Ok, if I understand you correctly then you may configure the board for four axes and use the [stepper spindle](https://github.com/grblHAL/Plugins_spindle/blob/master/stepper.c) to run the motor - it will claim the fourth (A) axis. Or do you want to add outputs for an extra axis and use the onboard drivers for a XYYZ configuration?
FYI if you double up the Z-axis you cannot control the two motors independently without hacking the code.

**Reply by Bazmundi** [reply]
Assuming the board config is currently for the out of the box EXYZ1Z2 per the [manual](https://github.com/bigtreetech/SKR-Pico/blob/master/BTT%20SKR%20Pico%20V1.0%20Instruction%20Manual.pdf), yes I am looking at swapping Y and Z in the map file to make the board behave as EXZY1Y2, with Y1Y2 driving the two steppers on the Y axis of the laser scanner rig. 

The z axis I am adding will be a [single stepper linear actuator](https://www.aliexpress.com/item/1005007051183598.html?spm=a2g0o.detail.pcDetailTopMoreOtherSeller.2.310fiinaiina8w&gps-id=pcDetailTopMoreOtherSeller&scm=1007.40000.327270.0&scm_id=1007.40000.327270.0&scm-url=1007.40000.327270.0&pvid=209dd462-83a7-4de2-b49d-2972cd5c46a8&_t=gps-id:pcDetailTopMoreOtherSeller,scm-url:1007.40000.327270.0,pvid:209dd462-83a7-4de2-b49d-2972cd5c46a8,tpp_buckets:668%232846%238114%231999&pdp_npi=4%40dis%21AUD%2126.80%2126.80%21%21%2117.73%2117.73%21%402101e54417245584553764181e4e56%2112000039234090356%21rec%21AU%21117209436%21X&utparam-url=scene%3ApcDetailTopMoreOtherSeller%7Cquery_from%3A).

**Reply by terjeio** [reply]
> Assuming the board config is currently for the out of the box EXYZ1Z2

It is not, the manual is for 3D printer setups. Default for grblHAL is XYZ. E is not supported and the motor driver can be used for a ganged/auto-squared axis or A-axis. The second Z connector on the board is just wired in parallel to first. So build grblHAL for four axes and connect the linear actuator to E (A-axis), both Y motors to Y and leave the second Z unconnected.

**Reply by Bazmundi** [reply]
So you're saying what?

```
// Define step pulse output pins.
#define STEP_PORT                   GPIO_PIO_1 // Single pin PIO SM
#define X_STEP_PIN                  11 // is not X of EXYZ1Z2 per the BTT SKR PICO manual
#define Y_STEP_PIN                  6   // is not Y of EXYZ1Z2 per the BTT SKR PICO manual
#define Z_STEP_PIN                  19 // is not Zx of EXYZ1Z2 per the BTT SKR PICO manual

etc.

```

So you're saying use something like the following from the webbuilder:

```
{"driver":"RP2040","URL":"https://github.com/grblHAL/RP2040","board":"BOARD_BTT_SKR_PICO_10","symbols":["OVERRIDE_MY_MACHINE=1","USB_SERIAL_CDC=1","N_AXIS=4","PROBE_ENABLE=0","TRINAMIC_ENABLE=2209","TRINAMIC_UART_ENABLE=1","ESTOP_ENABLE=1","SPINDLE0_ENABLE=7","N_SPINDLE=1"],"docker_instance":"/RP2040","grblHAL":"webBuilder"}
```

Then with the webbuilder approach, wire up the new Z axis to what the BTT SKR PICO manual calls the E (aka A-axis),  use a "Y" loom to wire up to both "Y" motors on the laser cutter to what the BTT SKR PICO manual calls the Y axis, so where is grbl X connected to PICO X or PICO Zx.

Sounds like even that is making it too hard.  I was looking at using the Z1Z2 as Y1Y2 simply to use standard cables from the steppers, seemed a clean way to avoid hacking non-standard cables.   If the XYZ of grbl is to XY(Z1Z2) of the BTT SKR PICO, and I use a "Y" loom as you suggested, is not the answer just to do a standard 3 axis build, connect grbl X to PICO X, grbl Y to PICO Y (with the "Y" loom and so to both Y motors on the laser cutter) and then just use the PICO Z1 as the grbl Z (and, as you say, don't connect to Z2)?

**Reply by terjeio** [reply]
> Sounds like even that is making it too hard. I was looking at using the Z1Z2 as Y1Y2 simply to use standard cables from the steppers, seemed a clean way to avoid hacking non-standard cables.

If you find it easier to swap pin mappings for Y and Z and compile yourself then go for it.

**Reply by Bazmundi** [reply]
Yep, easier if I can get the environment setup :smiley:, still bumbling through that.  Thanks for the ideas though.  If I give up on trying to get the build working, hacking up a cable looks like an option.

**Reply by terjeio** [reply]
If your setup is XYYZ then you can use the Web Builder, the second Y motor is then assigned to E...

**Reply by Bazmundi** [reply]
Ah ha! That was what I was not getting. So is that just by setting to 4 axis?  And Spindle 1 to ON/OFF and I am done!

Or is it 3 axis and Y made Ganged or Auto-Squared?  You suggested earlier it my be either, but which would you recommend?  I'll try ganged and see how that goes.  That still drives E right? Just no fancy alignment correction.  That should do for a solder paste dispenser I think.

Thanks very much for making that clearer! I am trying to do a crash course on all of this.

**Reply by terjeio** [reply, marked as answer]
> Or is it 3 axis and Y made Ganged or Auto-Squared? Y

This.

**Reply by Bazmundi** [reply]
Much appreciate your help! 

I will still pour over the code to see how it works, but the webbuilder is GOLD!
